tools/build_langserver/langserver/fs.go

```diff
@@ -0,0 +1,74 @@
+package langserver
+
+import (
+	"context"
+	"encoding/json"
+
+	"tools/build_langserver/lsp"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+func (h *LsHandler) handleFSRequests(ctx context.Context, req *jsonrpc2.Request) (result interface{}, err error) {
+	if !isFileSystemMethod(req) {
+		return nil, nil
+	}
+
+	log.Info("Handling fs method %s, with param %s", req.Method, req.Params)
+
+	switch req.Method {
+	case "textDocument/didOpen":
+		var params lsp.DidOpenTextDocumentParams
+		if err := json.Unmarshal(*req.Params, &params); err != nil {
+			return nil, err
+		}
+		h.workspace.Store(params.TextDocument.URI, params.TextDocument.Text)
+		return nil, nil
+	case "textDocument/didChange":
+		var params lsp.DidChangeTextDocumentParams
+		if err := json.Unmarshal(*req.Params, &params); err != nil {
+			return nil, err
+		}
+
+		err := h.workspace.TrackEdit(params.TextDocument.URI, params.ContentChanges)
+		if err != nil {
+			return nil, err
+		}
+		return nil, nil
+	case "textDocument/didSave":
+		var params lsp.DidSaveTextDocumentParams
+		if err := json.Unmarshal(*req.Params, &params); err != nil {
+			return nil, err
+		}
+		err := h.workspace.Update(params.TextDocument.URI, params.Text)
+		if err != nil {
+			return nil, err
+		}
+		return nil, nil
+	case "textDocument/didClose":
+		var params lsp.DidCloseTextDocumentParams
+		if err := json.Unmarshal(*req.Params, &params); err != nil {
+			return nil, err
+		}
+		delete(h.workspace.documents, params.TextDocument.URI)
```

Comment:
I would probably put something like this into `h.workspace` as a function, even if it's simple, so it manages its internal data structures. Will make it easier if you need to keep an eye on it (e.g. you might want to introduce locking around map accesses for thread-safety).

Reply:
That's a good idea actually, it would be more consistent to the other workspace methods as well